Add support for F# methods that return Async<T> or Task/Task<T>.

petejohanson opened the pull request:
With these changes, the following types of F# test functions seem to work as expected:

1. Returning native F# async (the signature here would be `unit -> Async<unit>`):

```
let ``it should async return the expected results``() =
    async {
        let! result = someAsyncMethod "test"
        result.ShouldBe("testtest")
    }
```

2. Or the version that returns a `Task<T>`, but doesn't have the attribute that's added for `async` C# methods (the signature here would be `unit -> Task<unit>`):

```
let ``it should async return the expected results``() =
    async {
        let! result = someAsyncMethod "test"
        result.ShouldBe("testtest")
    } |> Async.StartAsTask
```

I'm not honestly sure how to easily test this without dragging F# into the tests, or writing a second test assembly in F#.

Thoughts?

plioi commented:
At a glance, this looks great. I like how the change conforms to some existing patterns, and I don't expect to have any objections.  To set expectations, I likely won't get to really try this out until the weekend.

plioi commented:
I'll be able to set up some end-to-end tests with F# over in the companion repo, https://github.com/fixie/fixie.integration

petejohanson commented:
Sounds good. Thanks!

plioi commented:
Upon closer inspection, I see a few things worth adjusting.

1. This should be submitted against the 'develop' branch (whose content is nearly identical to master right now). Just now, I corrected the github repo settings so that that will be the default branch for PRs from now on.
2. In the call to MethodBase.Invoke(...), the final null argument is highlighted by ReSharper as being suspicious: "Possible 'null' assignment to entity marked with 'NotNull' attribute." Further, documentation on the parameters for this Invoke overload say that when the third argument (binder) is null, the second argument (BindingFlags.Public | BindingFlags.Static) is ignored. I think what we really want is to call the other overload of Invoke. Passing in null for the first argument to *that* overload is appropriate for a static method like this.
3. The hard cast to (Task) near the end could throw if the test method does the admittedly-silly thing of ultimately returning null. I think that's a lurking preexisting bug for C# async tests, too!
4. We're mixing compile-time reflection information with runtime type information, and might get surprised in some situations. `IsAsync` here is defined with an `== typeof(Task)` comparison, which will be true for `Task` but not for `Task<T>`. We might want to move all these runtime concerns from ReflectionExtensions to private members within MethodInfoExtensions.

petejohanson commented:
@plioi Thanks for the feedback, I addressed most of the items you listed, including tweaking `IsAsync` to handle both `Task` and `Task<T>` responses, even when not returned from `async` decorated methods.

I added some more scenarios to the `AsyncCaseTests` as well, to help exercise some of this. I'm still not clear what exactly you were proposing for the ReflectionExtensions versus MethodInfoExtension uses. Can you elaborate, and I'm happy to tweak in this PR?

Thanks.

plioi commented:
@petejohanson I've released this in Fixie 2.0.3.

plioi commented:
https://github.com/fixie/fixie.integration/blob/master/src/FSharp.Tests/AsyncTests.fs This sample test file demonstrates the new behavior. Before this fix, all 4 would appear as if they passed.

petejohanson commented:
Working like a champ here in my project. Thanks!
